### PR TITLE
[next] Allow writing components using Typescript and provide type definitions for `NcButton`

### PR DESCRIPTION
**Pull request by susnux**
### ☑️ Resolves

* Start of #841 

Add the Typescript infrastructure to write components using Typescript and migrate a first example component (`NcButton`) to Typescript. The type definitions are generated and published.

### Todo
As discussed previously the legacy naming of our entry points (`components` vs `Components`) will lead to a crowded `dist` directory, as Typescript uses the source naming, but our entry points use the upper case naming.
Maybe we should unify those names, meaning either rename the source folder or the entry points (?)

### 🏁 Checklist

- [x] ⛑️ Tests are included or are not applicable
- [x] 📘 Component documentation has been extended, updated or is not applicable

### CODE REVIEW

File: src/components/NcButton/NcButton.vue [hunk 1]
```diff
@@ -554,7 +565,7 @@ export default {
 
 		// If we have a router-link, we wrap the button in it
 		if (this.to) {
-			return h(resolveComponent('router-link'), {
+			return h(RouterLink, {
```
Comment (ShGKme) [inline src/components/NcButton/NcButton.vue:568]: **issue (blocking):** `vue-router` is not a dependency or peerDependency of `@nextcloud/vue`

File: package.json [hunk 2]
```diff
@@ -110,6 +111,7 @@
     "@vue/vue3-jest": "^29.2.5",
     "babel-jest": "^29.0.1",
     "babel-loader-exclude-node-modules-except": "^1.2.1",
+    "babel-preset-typescript-vue3": "^2.0.17",
```
Comment (ShGKme) [inline package.json:114]: This preset seems to not be very widely-known and maintained with almost no installations, 5 stars, and rare updates...

**Comment by susnux** [comment]
BTW currently the vue3 eslint config is a bit outdated, so there is no typescript config as for the vue2 config. So eslint fails currently.

File: src/components/NcButton/NcButton.vue [hunk 3]
```diff
@@ -321,24 +321,32 @@ export default {
 ```
 </docs>
 
-<script>
+<script lang="ts">
 import isSlotPopulated from '../../utils/isSlotPopulated.js'
```
Comment (raimund-schluessler) [inline src/components/NcButton/NcButton.vue:325]:
```suggestion
import isSlotPopulated from '../../utils/isSlotPopulated'
```

File: src/utils/isSlotPopulated.ts [hunk 4]
```diff
@@ -20,18 +20,22 @@
  *
  */
 
-import { Fragment, Comment, Text } from 'vue'
+import { Fragment, Comment, Text, type VNode, type VNodeNormalizedChildren } from 'vue'
```
[the remark this answers is not in the document]
Comment (susnux) [reply src/utils/isSlotPopulated.ts:23]: It is possible, but most of the time I write a extra types import to make it clearer what are types and what logic.

File: src/components/NcButton/NcButton.vue (hunk 3 above)
Comment (raimund-schluessler) [reply src/components/NcButton/NcButton.vue:325]: Well, I don't know, I just saw the error message from jest
```
 Cannot find module '../../utils/isSlotPopulated.js' from 'src/components/NcButton/NcButton.vue'
```

Comment (susnux) [reply src/components/NcButton/NcButton.vue]: Good point, that needs to be fixed

File: src/components/NcButton/NcButton.vue (hunk 1 above)
Comment (susnux) [reply src/components/NcButton/NcButton.vue]: Done

File: package.json (hunk 2 above)
Comment (susnux) [reply package.json:126]: Well the alternative would be copy-paste the rules from the preset, we only need this for vue styleguideist which still uses webpack+babel

File: src/components/NcButton/NcButton.vue (hunk 3 above)
Comment (susnux) [reply src/components/NcButton/NcButton.vue]: Solved :heavy_check_mark:

File: package.json (hunk 2 above)
Comment (raimund-schluessler) [reply package.json:127]: Or we get rid of vue-styleguidist and webpack first and migrate to vitepress? But I guess we can also keep it for now and migrate later on.

File: package.json [hunk 5]
```diff
@@ -109,21 +109,22 @@
...
     "@mdi/js": "^7.3.67",
     "@mdi/svg": "^7.0.96",
     "@nextcloud/babel-config": "^1.0.0",
     "@nextcloud/browserslist-config": "^3.0.0",
-    "@nextcloud/eslint-config": "github:nextcloud-libraries/eslint-config#fe67f6b84e3c13eb0ea2478f44939abc2fe7c99c",
+    "@nextcloud/eslint-config": "github:nextcloud-libraries/eslint-config#0d785eb6d1d62b91be1452759a615a476c87db4c",
     "@nextcloud/stylelint-config": "^2.3.1",
     "@nextcloud/vite-config": "github:nextcloud-libraries/nextcloud-vite-config#36ae8cf759a30e1b020d90a6c3585fa7b7734dc7",
     "@nextcloud/webpack-vue-config": "github:nextcloud/webpack-vue-config#1f4277c6ab14be63810f425ef6b762d8ca59697b",
+    "@vue/compiler-sfc": "^3.3.7",
```
Comment (raimund-schluessler) [inline package.json:123]: Do we really need to add this manually? The docs say, one doesn't:
https://vuejs.org/guide/scaling-up/tooling.html#vue-compiler-sfc

Installing it manually might break dependabot updates.

File: src/components/NcActionCheckbox/NcActionCheckbox.vue [hunk 6]
```diff
@@ -118,7 +118,7 @@ export default {
 	},
 
 	methods: {
-		checkInput(event) {
+		checkInput(/* event */) {
```
Comment (raimund-schluessler) [inline src/components/NcActionCheckbox/NcActionCheckbox.vue:121]: Is this necessary for typescript? Just asking out of curiosity.

File: src/components/NcSelect/NcSelect.vue [hunk 7]
```diff
@@ -983,15 +983,12 @@ export default {
...
-				userSelect,
-				// Props to forward
-				...initialPropsToForward
-			} = this.$props
+			const initialPropsToForward = { ...this.$props }
+			// Props handled by this component
+			delete initialPropsToForward.inputClass
+			delete initialPropsToForward.noWrap
+			delete initialPropsToForward.placement
+			delete initialPropsToForward.userSelect
```
Comment (raimund-schluessler) [inline src/components/NcSelect/NcSelect.vue:991]: This seems unrelated, no?

File: src/utils/isSlotPopulated.ts [hunk 8]
```diff
@@ -20,18 +20,23 @@
...
- * @param {Array} vnodes The array of vnodes to check
+ * @param vnodes The array of vnodes to check
  */
-const isSlotPopulated = function(vnodes) {
-	return !!vnodes?.some(node => {
-		if (node.type === Comment) return false
-		if (node.type === Fragment && !isSlotPopulated(node.children)) return false
-		if (node.type === Text && !node.children.trim()) return false
+const isSlotPopulated = function(vnodes?: VNode[] | VNodeNormalizedChildren) {
+	return Array.isArray(vnodes) && vnodes.some(node => {
```
Comment (raimund-schluessler) [inline src/utils/isSlotPopulated.ts:32]: Why do you prefer
```ts
Array.isArray(vnodes) && vnodes.some(...)
```
over
```js
!!vnodes?.some(...)
```
? Does it behave different in some cases?

File: package.json (hunk 2 above)
Comment (susnux) [reply package.json:127]: Or even migrate to #174

File: src/components/NcActionCheckbox/NcActionCheckbox.vue (hunk 6 above)
Comment (susnux) [reply src/components/NcActionCheckbox/NcActionCheckbox.vue:121]: I can also remove it, but it is unused and otherwise ESLint will complain about unused parameter

File: src/components/NcSelect/NcSelect.vue (hunk 7 above)
Comment (susnux) [reply src/components/NcSelect/NcSelect.vue:991]: No this fixes ESLint which will complain about unused variables, because you create const variables `inputClass`, `noWrap` etc without using them.
So instead of creating unused variables, just delete the props we use.

File: package.json (hunk 2 above)
Comment (raimund-schluessler) [reply package.json:127]: > Or even migrate to #174

Why not, but that looks like a lot more work than vitepress, no?

File: src/utils/isSlotPopulated.ts (hunk 8 above)
Comment (susnux) [reply src/utils/isSlotPopulated.ts:32]: Because vnodes is not always an array, you call the function like this `isSlotPopulated(node.children)` so the the parameter type for that call is `VNodeNormalizedChildren` which can be `null, string, VNode[], RawSlots`.

So when you want to call `some` on that parameter it will cause a Typescript error:

> Property 'some' does not exist on type 'string | VNode<RendererNode, RendererElement, { [key: string]: any; }>[] | VNodeArrayChildren | RawSlots'.
> Property 'some' does not exist on type 'string'.ts(2339)

So Array.isArray simply asserts that the parameter is an Array, so the following code will assume that vnodes can only be of type `VNode[]`.

File: src/components/NcActionCheckbox/NcActionCheckbox.vue (hunk 6 above)
Comment (susnux) [reply src/components/NcActionCheckbox/NcActionCheckbox.vue:121]: What would you prefer? Let it there as a comment or remove the unused argument?

Comment (raimund-schluessler) [reply src/components/NcActionCheckbox/NcActionCheckbox.vue:121]: I don't mind. You can leave it as a comment.

File: src/utils/isSlotPopulated.ts (hunk 8 above)
Comment (raimund-schluessler) [reply src/utils/isSlotPopulated.ts:32]: Indeed, that's what I had in mind. But checking that's an array first seems like the better solution here. 👍

File: package.json (hunk 2 above)
Comment (susnux) [reply package.json:126]: Yes that is what @ShGKme and I fear :sweat_smile: 
But it has some great benefits, as always the problem is missing time.

File: src/components/NcSelect/NcSelect.vue (hunk 7 above)
Comment (susnux) [reply src/components/NcSelect/NcSelect.vue:991]: (BTW I personally even think the explicit delete is more readable meaning it makes it clear what is expected to happen)

File: package.json (hunk 2 above)
Comment (ShGKme) [reply package.json:126]: Does Vitepress mean to write all the documentation manually? Event props list...

Comment (raimund-schluessler) [reply package.json:126]: > Does Vitepress mean to write all the documentation manually? Even props list...

No, this will be extracted. @susnux started a first try in #3565, see https://deploy-preview-3565--nextcloud-vue-components.netlify.app
It already kind of works, but needs some adjustements and a fix for the components preview.

Comment (susnux) [reply package.json:126]: Yes instead of using vue-styleguideist which uses webpack / vuepress we can use the vue-styleguidist cli tools to extract the documentation as markdown which we then serve using vitepress, see the example:

https://github.com/vue-styleguidist/vue-styleguidist/tree/dev/examples/docgen-vite

Comment (susnux) [reply package.json:128]: I dropped that preset and added a simple config.

**Comment by susnux** [comment]
Rebased and solved some issues after all components were migrated to Vue 3. So I would say this is ready for reviews :)

File: .eslintrc.js [hunk 9]
```diff
@@ -21,4 +21,17 @@ module.exports = {
 			],
 		},
 	},
+	rules: {
+		'jsdoc/check-tag-names': [
+			'warn',
+			{
+				definedTags: [
+					'values', // required for vue-styleguideist
```
Comment (ShGKme) [inline .eslintrc.js:29]: It is not required for styleguideist, it supports `@type` with string alternative `@type {'a'|'b'|'c'}`

Comment (ShGKme) [reply .eslintrc.js:29]: And we don't have `@values` anymore

**Review by skjnldsv — APPROVED** [review]

File: .eslintrc.js [hunk 10]
```diff
@@ -21,4 +21,17 @@ module.exports = {
...
+					'notExported',
+					'packageDocumentation',
+					'jest-environment',
```
Comment (raimund-schluessler) [inline .eslintrc.js:32]: Jest is not used anymore. Can we get rid of this then?

**Review by raimund-schluessler — APPROVED** [review]

File: .eslintrc.js (hunk 9 above)
Comment (susnux) [reply .eslintrc.js:29]: Fixed

File: .eslintrc.js (hunk 10 above)
Comment (susnux) [reply .eslintrc.js:32]: Yes, fixed :)

**Review by ShGKme — APPROVED** [review]
